### PR TITLE
Implement backpropagation for SVD when full_matrices is False

### PR DESCRIPTION
**Pull request by vnavkal**
This pull request implements backpropagation for the singular value decomposition operation when `full_matrices` is False.  A request for this feature is in #13641.

The algorithm I use is the same one as in the [autograd implementation](https://github.com/HIPS/autograd/blob/77387abc6d5d57a2688d6fa04e932f4a9d5613d9/autograd/numpy/linalg.py#L127-L217).  There is a very nice derivation of the formula, written (I think) by @j-towns, at https://j-towns.github.io/papers/svd-derivative.pdf.  (@j-towns I hope you don't mind that I include this link here; I found it in the description of your autograd pull request that added this feature.)

Thanks in advance for taking the time to review this!

### CODE REVIEW

File: tensorflow/python/ops/linalg_grad.py [hunk 1]
```diff
@@ -334,9 +335,10 @@ def _SvdGrad(op, grad_s, grad_u, grad_v):
     # multiple singular values with value zero. I am not sure if this is a true
     # instability or if it simply throws off the finite difference gradient
     # checker.
```
Comment (vnavkal) [inline tensorflow/python/ops/linalg_grad.py]: How should I update this comment?  I personally feel there isn't much hope for the case `abs(m-n) > 1` when `full_matrices=True`, because the gradient would need to be consistent with the particular choices of U and V made by the SVD algorithm.  So my preference is to remove the comment.  But if others prefer for me to leave it here, I can instead change `numerically stable for abs(m-n) > 1` to `numerically stable for abs(m-n) > 1 and full_matrices=True`.

**Comment by j-towns** [comment]
>@j-towns I hope you don't mind that I include this link here; I found it in the description of your autograd pull request that added this feature.

No problem. Glad you found it useful!

**Comment by rmlarsen** [comment]
Will review this on the next couple of days. Thanks for adding this!

File: tensorflow/python/ops/linalg_grad.py [hunk 2]
```diff
@@ -352,32 +354,45 @@ def _SvdGrad(op, grad_s, grad_u, grad_v):
...
+
     u_gu = math_ops.matmul(u, grad_u, adjoint_a=True)
-    v_gv = math_ops.matmul(v, grad_v, adjoint_a=True)
+    v_gv = math_ops.matmul(v1, grad_v1, adjoint_a=True)
 
-    if m == n:
-      f_u = f * u_gu
-      f_v = f * v_gv
-    else:
-      dv2 = array_ops.matrix_transpose(v_gv[..., m:n, :m]) - v_gv[..., :m, m:n]
```
Comment (vnavkal) [inline tensorflow/python/ops/linalg_grad.py]: I should explain what happened to `dv2`.  `dv2` has two terms:
1. `array_ops.matrix_transpose(v_gv[..., m:n, :m])` and
2. `-v_gv[..., :m, m:n]`.

When `grad_a_nouv` gets right-multiplied by the adjoint of `v`, the submatrix `dv2` gets right-multiplied by the adjoint of `v[..., :, m:n]`.  After right-multiplication by `v[..., :, m:n]`,
1. the first term of `dv2` becomes the first term of `term2_nous`: `math_ops.matmul(grad_v1, proj_v1_perp, adjoint_a=True)`
2. the second term of `dv2` becomes the second term of `term2_nous`: `-math_ops.matmul(v1t_gv2, v2, adjoint_b=True)`

To see why the first terms match, use the identity
```
math_ops.matmul(v[..., :, :m], v[..., :, :m], adjoint_b=True)
   + math_ops.matmul(v[..., :, m:n], v[..., :, m:n], adjoint_b=True)
       == linalg_ops.eye(n, dtype=v.dtype)
```
which follows from the fact that `v` is unitary.

My motivation for separating the terms of `dv2` was to make clear what the contribution to `grad_a` is from the "redundant" columns of V when `full_matrices=True`.

If the math here isn't clear, just let me know and I can write it down and upload it.

**Comment by vnavkal** [comment]
sure, no hurry, thanks!

**Comment by martinwicke** [comment]
ping @rmlarsen

File: tensorflow/python/ops/linalg_grad.py (hunk 1 above)
Comment (rmlarsen) [reply tensorflow/python/ops/linalg_grad.py]: I'm OK with removing the comment.

**Comment by rmlarsen** [comment]
@tensorflow-jenkins Test this please

File: tensorflow/python/ops/linalg_grad.py (hunk 1 above)
Comment (rmlarsen) [reply tensorflow/python/ops/linalg_grad.py]: I'm OK with removing the comment.

File: tensorflow/python/ops/linalg_grad.py [hunk 3]
```diff
@@ -300,7 +301,7 @@ def _SvdGrad(op, grad_s, grad_u, grad_v):
         "SVD gradient has not been implemented for input with unknown "
         "inner matrix shape.")
 
-  if not op.get_attr("full_matrices") or not op.get_attr("compute_uv"):
+  if not op.get_attr("compute_uv"):
     s, u, v = linalg_ops.svd(a, compute_uv=True, full_matrices=True)
```
Comment (rmlarsen) [inline tensorflow/python/ops/linalg_grad.py]: Remove full_matrices=True here?

File: tensorflow/python/ops/linalg_grad.py [hunk 4]
```diff
@@ -352,32 +354,45 @@ def _SvdGrad(op, grad_s, grad_u, grad_v):
...
 
-    if m != n:
-      grad_a_nouv = array_ops.concat(
-          [grad_a_nouv, math_ops.matmul(s_inv_mat, dv2)], -1)
+    term1 = math_ops.matmul(u, math_ops.matmul(term1_nouv, v1, adjoint_b=True))
+
+    if m == n:
+      grad_a_before_transpose = term1
+    else:
+      proj_v1_perp = (linalg_ops.eye(n, dtype=v.dtype)
```
Comment (rmlarsen) [inline tensorflow/python/ops/linalg_grad.py]: proj_v1_perp is large potentially very large. I think it would be more efficient to compute term2_nous as

grad_v1 - math_ops.matmul(math_ops.matmul(grad_v1, v1), v1, adjoint_b=True)

**Review by rmlarsen — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/linalg_grad.py (hunk 4 above)
Comment (vnavkal) [reply tensorflow/python/ops/linalg_grad.py]: good point.  I updated this part of the calculation so that it doesn't use an `n` by `n` matrix.  (Note that in the formula you wrote, `grad_v1` needs to be replaced by its adjoint.)

File: tensorflow/python/ops/linalg_grad.py (hunk 3 above)
Comment (vnavkal) [reply tensorflow/python/ops/linalg_grad.py]: I kept it this way to preserve existing behavior; my pull request changes only the logic for the case when `compute_uv` is `True`.  However, I think you're right that it could be removed, and `v[..., :, :m]` could be replaced with `v` [here](https://github.com/tensorflow/tensorflow/blob/17e725c0558581cba19bd6c409698b2c3f88efe5/tensorflow/python/ops/linalg_grad.py#L324) and [here](https://github.com/tensorflow/tensorflow/blob/17e725c0558581cba19bd6c409698b2c3f88efe5/tensorflow/python/ops/linalg_grad.py#L328).

Would it make sense for me to make this change in a separate pull request, since it's not directly related to this one?  I can also just add it to this one if you prefer.

**Review by rmlarsen — APPROVED** [review]

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by helwilliams** [comment]
Hello I am trying to use SVD single components in my training process, however, occasionally the training I get the following error maybe due to back propagation of SVD being unstable, I am using compute_uv=False as I only care about s for training   
```
 InvalidArgumentError (see above for traceback): slice index 0 of dimension 0 out of bounds.
         [[node worker_0/strided_slice (defined at g:\NiftyNet\niftynet\layer\loss_added.py:61)  = StridedSlice[Index=DT_INT32, T=DT_FLOAT, begin_mask=0, ellipsis_mask=0, end_mask=0, new_axis_mask=0, shrink_axis_mask=1, _device="/job:localhost/replica:0/task:0/device:GPU:0"](worker_0/Svd, ConstantFolding/worker_0/gradients/worker_0/NoNewNet/R4/120_bn_relu/bn_/batch_norm/add_grad/BroadcastGradientArgs-folded-1, worker_0/strided_slice/stack_1, worker_0/strided_slice/stack_1)]]

```
any advice on how to bypass this problem would be great, thanks and thanks for your work on the tool.
